codec/src/main/java/io/netty/handler/codec/MessageAggregatorNew.java

```diff
@@ -0,0 +1,392 @@
+            return true;
+        }
+        return aggregating && tryContentMessage(msg) != null;
+    }
+
+    /**
+     * If the passed {@code msg} is a {@linkplain S start message} then cast and return the same, else return
+     * {@code null}.
+     */
+    protected abstract S tryStartMessage(Object msg);
+
+    /**
+     * If the passed {@code msg} is a {@linkplain C content message} then cast and return the same, else return
+     * {@code null}.
+     */
+    protected abstract C tryContentMessage(Object msg);
+
+    /**
+     * Returns {@code true} if and only if the specified message is the last content message. Typically, this method is
+     * implemented as a single {@code return} statement with {@code instanceof}:
+     * <pre>
+     * return msg instanceof MyLastContentMessage;
+     * </pre>
+     * or with {@code instanceof} and boolean field check:
+     * <pre>
+     * return msg instanceof MyContentMessage && msg.isLastFragment();
+     * </pre>
+     */
+    protected abstract boolean isLastContentMessage(C msg) throws Exception;
+
+    /**
+     * Returns {@code true} if and only if the specified message is already aggregated.  If this method returns
+     * {@code true}, this handler will simply forward the message to the next handler as-is.
+     */
+    protected abstract boolean isAggregated(Object msg) throws Exception;
+
+    /**
+     * Returns the length in bytes of the passed message.
+     *
+     * @param msg to calculate length.
+     * @return Length in bytes of the passed message.
+     */
+    protected abstract int lengthForContent(C msg);
+
+    /**
+     * Returns the length in bytes of the passed message.
+     *
+     * @param msg to calculate length.
+     * @return Length in bytes of the passed message.
+     */
+    protected abstract int lengthForAggregation(A msg);
+
+    /**
+     * Returns the maximum allowed length of the aggregated message in bytes.
+     */
+    public final int maxContentLength() {
+        return maxContentLength;
+    }
+
+    protected final ChannelHandlerContext ctx() {
+        if (ctx == null) {
+            throw new IllegalStateException("not added to a pipeline yet");
+        }
+        return ctx;
+    }
+
+    @Override
+    protected void decode(final ChannelHandlerContext ctx, I msg) throws Exception {
+        assert aggregating;
+        final S startMsg = tryStartMessage(msg);
+        if (startMsg != null) {
+            handlingOversizedMessage = false;
+            if (currentMessage != null) {
+                currentMessage.close();
+                currentMessage = null;
+                throw new MessageAggregationException();
+            }
+
+            // Send the continue response if necessary (e.g. 'Expect: 100-continue' header)
+            // Check before content length. Failing an expectation may result in a different response being sent.
+            Object continueResponse = newContinueResponse(startMsg, maxContentLength, ctx.pipeline());
```

Comment:
This sounds HTTP specific while the name and placement of this class sounds wholly protocol agnostic.

Reply:
oh yeah welcome to the weird class 🤣 ... this is basically a port of existing `MessageAggregator` to use `Buffer`.